[Lens] Stack as percentage

flash1293 opened the pull request:
Related: https://github.com/elastic/kibana/issues/57389
Fixes: https://github.com/elastic/kibana/issues/72320

This PR implements a "percentage mode" for area and bar charts. There are quite some uncertainties around this I want to explore with this PR.

<img width="669" alt="Screenshot 2020-07-03 at 13 38 18" src="https://user-images.githubusercontent.com/1508364/86465722-8116f000-bd32-11ea-8e55-2d9bfa0e5099.png">

TODOs

* [x] Write tests
* [x] Integrate additional icons

## Functionality

There are multiple options how to offer this functionality - it could be a flag in the chart level settings, a layer level setting or even a metric level setting. However I think this is best shown as a new "chart type" in the popover:
<img width="393" alt="Screenshot 2020-07-03 at 13 38 23" src="https://user-images.githubusercontent.com/1508364/86465827-b1f72500-bd32-11ea-9618-271e261afeac.png">

We would need an icon similar to the "stacked" chart icons.

This also includes a suggestion to show the current chart as percentage if a stacked chart is used:

<img width="400" alt="Screenshot 2020-07-03 at 13 45 14" src="https://user-images.githubusercontent.com/1508364/86466236-92143100-bd33-11ea-825b-ce17430a386d.png">

It's possible to use a percentage layer along with a non-percentage layer:
<img width="958" alt="Screenshot 2020-07-03 at 13 57 44" src="https://user-images.githubusercontent.com/1508364/86467230-537f7600-bd35-11ea-98db-24d99a53f3a9.png">

Transitions between charts are described in https://github.com/elastic/kibana/pull/70703#issuecomment-662593173

This PR also changes the default percentage formatting to maximum two decimal places (instead of always showing three)

## Implementation

On a technical level these are simply new series types of the xy chart: `'bar_percentage_stacked'  | 'bar_horizontal_percentage_stacked' | 'area_percentage_stacked'`. If the type includes `percentage`, percentage mode will be activated on the series spec. Also, the axis formatter will be overwritten with a percentage formatter.

This PR does not implement the "Other" bucket for the terms operation - I get why this is related but I think there are still a lot of valid use cases for a percentage mode even without this feature. IMHO we should split it out into a separate PR.

flash1293 commented:
@wylieconlon @mbondyra @cchaos @AlonaNadler Looking for feedback on this one as well.

flash1293 commented:
@elasticmachine merge upstream

AlonaNadler commented:
The pr failed for me, anything Im missing?

flash1293 commented:
@AlonaNadler updated the PR, should work again.

mbondyra commented:
@flash1293 when it comes to how to offer the functionality, I agree with you, for me the most friendly would be to present it as another chart type (until we have 30 types of charts I think we're safe with going this direction and I don't think we plan many more soon). 
Same for 'others' - that could be added later. 
I've tested the P

flash1293 commented:
@mbondyra Thanks for the review, I didn't expect this failure so it's good to know :) I will make sure I get this one when cleaning up the PR.

I remember some early mock ups with a search bar and filters in the chart switcher -  if it gets too busy we can introduce something like this.

cchaos commented:
**Question**: What makes the configuration of a Stacked vs Percentage stacked chart so different that it needs to be a separate chart type?

flash1293 commented:
@cchaos It's not about difference, more about discoverability and consistent ui - we are treating stacked charts the same way and you could ask the same question there. None of the xy chart variants "need" to be a different chart type, we could also present all of these settings as a set of UI controls like it's done in Visualize:
<img width="478" alt="Screenshot 2020-07-16 at 18 02 27" src="https://user-images.githubusercontent.com/1508364/87694458-87738600-c78e-11ea-8469-af564b368e72.png">

cchaos commented:
I do think there is an added ease-of-use to supplying this chart as a specific chart type, but I also want to ensure that users who may start in a normal bar or stacked bar can easily change it (or know how to change it) to display as percentages. So a couple thoughts on that (and these might already be addressed, just want to reiterate them):

1. Switching to the percentage style from **any other x/y chart** type will be able to take **all** configurations without any problems
2. Perhaps we still offer this setting in the config popover for all x/y charts, so it's **another** place where they can find it

As for the chart type picker and the icons, we are starting to amass quite a bit of different x/y charts in particular. I can create some icons to indicate these new chart types, but I think we need to go further and:

1. Stop using the glyph style icons in the **layer settings**. Maintaining two different icon sets will get unruly and there's quite a disconnect between the two styles that can start creating confusion 
1. In the chart picker, start **categorizing** the chart types (maybe adding descriptions) and add a quick search/filter input.

flash1293 commented:
@cchaos Thanks for the feedback.

> Switching to the percentage style from any other x/y chart type will be able to take all configurations without any problems

I don't see any issues here as soon as Wylies bugfix for multi layer chart switches is solved. Technically this pretty similar to stacking.

> Perhaps we still offer this setting in the config popover for all x/y charts, so it's another place where they can find it

You mean the newly introduced "Settings" popover? We can add an option in there as well to convert all of the layers to percentage mode, but take into account we can also configure charts with only partial percentage mode ("Mixed XY chart")

> Stop using the glyph style icons in the layer settings. Maintaining two different icon sets will get unruly and there's quite a disconnect between the two styles that can start creating confusion
> In the chart picker, start categorizing the chart types (maybe adding descriptions) and add a quick search/filter input.

Both of these sound great, could you throw together a quick mock up? Then I can adjust the chart picker on this PR

cchaos commented:
> You mean the newly introduced "Settings" popover? 

No I mean the config popover, the one where you select the aggregation and field.

I can get you a mock next week

flash1293 commented:
> No I mean the config popover, the one where you select the aggregation and field.

Oh, that's a change in how the percentage mode is applied - right now it happens on a per layer basis, but putting it in there would apply it's a per metric setting.

cchaos commented:
I'm playing with the PR and trying to understand percentages better. 

From my understanding a percentage chart doesn't make sense to have multiple y-axis unless they are on separate layers. See this example:

<img width="976" alt="Screen Shot 2020-07-20 at 17 24 09 PM" src="https://user-images.githubusercontent.com/549577/87988361-2f01f880-caae-11ea-89b8-28e8845d981e.png">

So should we restrict percentage charts to a single y-axis per layer? If so, then it seems that this setting IS metric-based and not necessarily layer based. But we'd have to have some crazy logic about allowing the selection the metric config popover only if there's a single y-axis. We could possibly revisit this option later and concentrate it on just being a chart type for now. 

Is there also a short name for it? It's quite a mouthful

flash1293 commented:
@cchaos I agree in most cases a percentage chart with multiple metrics on y axis doesn't make sense, but it can make sense if the two metrics are in relation to each other, e.g. `netPrice`, `tax` - it's a difficult balance to make it hard to do the wrong thing without restricting justified use cases too much.

About the naming - the only alternative I can think of would be "ratio chart", but it might get confusing considering a quick function named "Filter ratio" or something along those lines. Do we even need to put the full name in the chart switcher like this?

AlonaNadler commented:
Feedback so far. I admit it was hard to test and kept failing on me, changes to intervals, and number of terms in the breakdown by didnt seem to take effect. 

* it makes sense to have a stacked percentage chart as a chart in the chart switchers.
  * We need new icons for that
  * I would organize the charts by types and similarities: first-row bar, stacked bar, stacked % bar. Second: area chart, stacked area, stacked % area. third: line, table, metric. fourth row: pie, donut treemap. last horizontal, horizontal stacked, horizontal percentage. Right now everything looks too similar in the first 3 rows of charts. also horizontal is not being used frequently.
 * stacked requires x-axis, y-axis, breakdown by. 
* Transitions: this is how I suggest we do the transitions 
![image](https://user-images.githubusercontent.com/32550316/88209974-4ead3400-cc08-11ea-8e89-aec6ff359fdb.png)




From empty state - if users started by selecting one of the stack % charts: 
  * if there is a x-axis and breakdown by but no y-axis we can default to count of records
   * if the chart missing a breakdown by - let's show the red error `required dimension`
    * if we have a time series with only x-axis, y-axis and users select to move to stacked percentage. How about we move the x-axis

flash1293 commented:
@AlonaNadler Fixed the bug with the area percentage chart. I couldn't reproduce the problem with the unchangeable terms and date histogram params, could you share steps to reproduce for that?

About the transitions: I like those in general, but we don't have a way to represent "timestamp (on entire time range)" - this is both a problem for building the suggestion as well as configuring it in general (as there is no "entire time range" setting, and using the custom interval comes with its own problems). Also, simply from a UX standpoint - will the user understand how to change that X axis to something meaningful? Not saying it's impossible but it would be a lot of effort to make that work smoothly, so I wonder whether there is a better way? As it's the least effort in the current system I would recommend using auto interval for that date histogram for now to keep the effort for this feature low and make a note for this when we get to improving suggestions in general.

flash1293 commented:
Discussed this with @AlonaNadler and the best option to handle transitions with percentage mode charts is to allow an empty x axis dimension in xy charts in general and behave like Visualize by rendering just a single data point. Percentage mode chart maps bucket dimension to break down first to always render meaningful percentage.

flash1293 commented:
I updated the branch, fixed tests and types, now the chart switch change and icons are pending.

@AlonaNadler If you want you can check out how the transitions work on this PR.

AlonaNadler commented:
Amazing work 👍 👏 
Transition to percentage chart is smooth, excellent !!

The transition back to pie chart/donut and treemap clears the chart

Overall really great work

flash1293 commented:
@AlonaNadler 

> The transition back to pie chart/donut and treemap clears the chart

What case exactly do you mean? If you don't do any changes, going back and forth should just work:
![percentage_transition](https://user-images.githubusercontent.com/1508364/90626416-3c301580-e21b-11ea-8d10-33be4d7e4cc0.gif)

kibanamachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69050/)
* Commit: 57f82a9a16f2e090d782afda70869ad125e92036
* Flaky suites:
  * oss-pluginFunctional

---

### [Test Failures](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69050/testReport)

<details><summary>Plugin Functional Tests.test/plugin_functional/test_suites/core/route·ts.core route timeouts idle socket should timeout if servers response is too slow</summary>

[Link to Jenkins](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69050/testReport//junit/Plugin%20Functional%20Tests/test_plugin_functional_test_suites_core_route·ts/core_route_timeouts_idle_socket_should_timeout_if_servers_response_is_too_slow)


#### Standard Out
```
Failed Tests Reporter:
  - Test has not failed recently on tracked branches

[00:00:00]       │
[00:00:00]         └-: core
[00:00:00]           └-> "before all" hook
[00:00:00]           └-: route
[00:00:00]             └-> "before all" hook
[00:00:00]             └-: timeouts
[00:00:00]               └-> "before all" hook
[00:00:00]               └-: idle socket
[00:00:00]                 └-> "before all" hook
[00:00:00]                 └-> should timeout if payload sending has too long of an idle period
[00:00:00]                   └-> "before each" hook: global before each
[00:00:00]                   └- ✓ pass  (39ms) "core route timeouts idle socket should timeout if payload sending has too long of an idle period"
[00:00:00]                 └-> should not timeout if payload sending does not have too long of an idle period
[00:00:00]                   └-> "before each" hook: global before each
[00:00:00]                   └- ✓ pass  (220ms) "core route timeouts idle socket should not timeout if payload sending does not have too long of an idle period"
[00:00:00]                 └-> should timeout if servers response is too slow
[00:00:00]                   └-> "before each" hook: global before each
[00:00:00]                   │ info Taking screenshot "/dev/shm/workspace/parallel/2/kibana/test/functional/screenshots/failure/core route timeouts idle socket should timeout if servers response is too slow.png"
[00:00:01]                   │ info Current URL is: data:/,
[00:00:01]                   │ info Saving page source to: /dev/shm/workspace/parallel/2/kibana/test/plugin_functional/failure_debug/html/core route timeouts idle socket should timeout if servers response is too slow.html
[00:00:01]                   └- ✖ fail: core route timeouts idle socket should timeout if servers response is too slow
[00:00:01]                   │      Error: expected 'read ECONNRESET' to equal 'socket hang up'
[00:00:01]                   │       at Assertion.assert (packages/kbn-expect/expect.js:100:11)
[00:00:01]                   │       at Assertion.be.Assertion.equal (packages/kbn-expect/expect.js:227:8)
[00:00:01]                   │       at Assertion.be (packages/kbn-expect/expect.js:69:22)
[00:00:01]                   │       at result.then.err (test/plugin_functional/test_suites/core/route.ts:147:38)
[00:00:01]                   │       at process._tickCallback (internal/process/next_tick.js:68:7)
[00:00:01]                   │ 
[00:00:01]                   │ 

```

#### Stack Trace
```
Error: expected 'read ECONNRESET' to equal 'socket hang up'
    at Assertion.assert (packages/kbn-expect/expect.js:100:11)
    at Assertion.be.Assertion.equal (packages/kbn-expect/expect.js:227:8)
    at Assertion.be (packages/kbn-expect/expect.js:69:22)
    at result.then.err (test/plugin_functional/test_suites/core/route.ts:147:38)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```
</details>

---

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/57f82a9a16f2e090d782afda70869ad125e92036) | diff | [baseline](https://github.com/elastic/kibana/commit/cebcc3a2295e6a4eb02e790114c18ce4790cbe25) |
| --- | --- | --- | --- |
| lens | 256 | +3 | 253 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/57f82a9a16f2e090d782afda70869ad125e92036) | diff | [baseline](https://github.com/elastic/kibana/commit/cebcc3a2295e6a4eb02e790114c18ce4790cbe25) |
| --- | --- | --- | --- |
| lens | 869.3KB | +15.5KB | 853.7KB |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"UNSTABLE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69050/","number":"69050","commit":"57f82a9a16f2e090d782afda70869ad125e92036"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

dej611 commented on x-pack/plugins/lens/public/xy_visualization/to_expression.ts:
```diff
@@ -79,8 +79,10 @@ export const buildExpression = (
-    Boolean(layer.xAccessor && layer.accessors.length)
+  const validLayers = state.layers.filter(
+    (layer): layer is ValidLayer =>
+      !(layer.seriesType.includes('percentage') && !layer.splitAccessor) &&
```
This just a stylistic suggestion, as that double negation was tricking me while reading:

```js
!layer.seriesType.includes('percentage') && 
  layer.splitAccessor && 
  Boolean(layer.accessors.length)
```

dej611 commented on x-pack/plugins/lens/public/xy_visualization/xy_expression.tsx:
```diff
@@ -423,7 +422,7 @@ export function XYChart({
         showGridLines={gridlinesVisibilitySettings?.x}
         gridLineStyle={{ strokeWidth: 2 }}
-        hide={filteredLayers[0].hide}
+        hide={filteredLayers[0].hide || !filteredLayers[0].xAccessor || filteredLayers[0].hide}
```
`filteredLayers[0].hide` is repeated twice here

dej611 reviewed (commented):
Code looks ok for me. Just two minor things (one of which is just a subjective stylistic thing). Will test locally now

AlonaNadler commented:
> I went back and forth on the default percentages to show for pie charts with @AlonaNadler: currently we use 3 decimal places, and here you are showing no decimal places. Can @AlonaNadler pick one and we'll use it for both charts?

Can we use two decimal point as the maximum by default?

flash1293 commented:
@AlonaNadler @wylieconlon 

* Addressed the review comments
* Changed formatting to maximum 2 decimal places. Also adjusted the label in the pie popover 
<img width="432" alt="Screenshot 2020-09-07 at 12 04 37" src="https://user-images.githubusercontent.com/1508364/92376892-8feb9b80-f103-11ea-9718-b3e8beb3f630.png">

@cchaos Could you check and approve the scss changes? Pretty sure they are inherited your PR on top of this branch anyway.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/)
* Commit: d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9) | diff | [baseline](https://github.com/elastic/kibana/commit/19bf9ed072ca15157e11671745e3783efe27d925) |
| --- | --- | --- | --- |
| lens | 258 | +3 | 255 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9) | diff | [baseline](https://github.com/elastic/kibana/commit/19bf9ed072ca15157e11671745e3783efe27d925) |
| --- | --- | --- | --- |
| lens | 890.3KB | +16.4KB | 873.9KB |

### History
* :broken_heart: [Build #69801](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/) failed 875d1d7cd052e64f711fd0b031936c4fbd760986
* :green_heart: [Build #69692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/) succeeded fe97375c19d92b02f98e5343901c12cdda1d5299
* :broken_heart: [Build #69681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69681/) failed 9b5fda21b5416dad6ea8ae2d1b294a3821b66e3c
* :broken_heart: [Build #69680](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69680/) failed cfe907bce365bb315959cdb53b8b0041507b89e4
* :green_heart: [Build #69656](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69656/) succeeded 95946bc08e6676eceacbffbf928396433a8f256a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69656/","number":"69656","commit":"95946bc08e6676eceacbffbf928396433a8f256a"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69680/","number":"69680","commit":"cfe907bce365bb315959cdb53b8b0041507b89e4"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69681/","number":"69681","commit":"9b5fda21b5416dad6ea8ae2d1b294a3821b66e3c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/","number":"69692","commit":"fe97375c19d92b02f98e5343901c12cdda1d5299"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/","number":"69801","commit":"875d1d7cd052e64f711fd0b031936c4fbd760986"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/","number":"72886","commit":"d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9"}]}
PIPELINE-->

wylieconlon reviewed (changes requested):
It's getting very close, but please update the formatters

cchaos commented on x-pack/plugins/lens/public/pie_visualization/toolbar.tsx:
```diff
@@ -187,7 +187,7 @@ export function PieToolbar(props: VisualizationToolbarProps<PieVisualizationStat
+              defaultMessage: 'Maximum decimal places for percent',
             })}
             fullWidth
             display="columnCompressed"
```
The label is really long and just isn't fitting well in this space as the column display. Let's just change this to  
```suggestion
            display="rowCompressed"
```

cchaos commented on x-pack/plugins/lens/public/xy_visualization/axes_configuration.ts:
```diff
@@ -49,9 +49,16 @@ export function getAxesConfiguration(
+        formatter = {
+          id: 'percent',
+          params: {
+            pattern: '0.[00]%',
```
I would also like to point out that the current slider for the decimals is really laggy and may need some sort of debounce. https://share.getcloudapp.com/rRu7AK6N

wylieconlon replied:
@cchaos that looks like an existing issue for pie charts- this is the XY chart where there's no slider

flash1293 replied:
It's configurable by configuring the percentage format in the metric dimension. I think it's a good approach for now, if changing the format will become super common we can think about exposing it more prominently.

@cchaos I will create a separate issue for the slider

cchaos commented on x-pack/plugins/lens/public/xy_visualization/state_helpers.ts:
```diff
@@ -8,7 +8,11 @@ import { EuiIconType } from '@elastic/eui/src/components/icon/icon';
 import { SeriesType, visualizationTypes, LayerConfig, YConfig } from './types';
 
 export function isHorizontalSeries(seriesType: SeriesType) {
```
I just noticed this for the first time, but I realized that you group horizontal series into it's own mixed chart type if there's multiple layers, but your using the Mixed XY chart icon. 

<img width="350" alt="Screen Shot 2020-09-08 at 16 18 40 PM" src="https://user-images.githubusercontent.com/549577/92525356-b0921f00-f1f1-11ea-8e87-809fb2b03616.png">


Can you change it to just display the Horizontal stacked bar chart icon?

cchaos commented on x-pack/plugins/lens/public/xy_visualization/types.ts:
```diff
@@ -344,57 +358,81 @@ export const visualizationTypes: VisualizationType[] = [
-    label: i18n.translate('xpack.lens.xyVisualization.lineLabel', {
-      defaultMessage: 'Line',
+    id: 'bar_horizontal_percentage_stacked',
+    icon: 'visBarHorizontalStacked',
+    largeIcon: LensIconChartBarHorizontalPercentage,
```
We can honestly also start getting rid of the distinction between `icon` and `largeIcon`. They should always display the custom one.

flash1293 replied:
Done, removed the `largeIcon` everywhere and used the custom icon for `icon` in all places.

cchaos reviewed (commented):
Just found a couple things when playing with the PR but overall seems good to me. I can approve after the few changes mentioned.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/)
* Commit: c2555e20ef6741cbc3f551d67472d93cf43525c8

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/c2555e20ef6741cbc3f551d67472d93cf43525c8) | diff | [baseline](https://github.com/elastic/kibana/commit/b6731b42d1ab94eb7c4f6f76c1f4be4b3df4b649) |
| --- | --- | --- | --- |
| lens | 258 | +3 | 255 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/c2555e20ef6741cbc3f551d67472d93cf43525c8) | diff | [baseline](https://github.com/elastic/kibana/commit/b6731b42d1ab94eb7c4f6f76c1f4be4b3df4b649) |
| --- | --- | --- | --- |
| lens | 889.9KB | +16.0KB | 874.0KB |

### History
* :green_heart: [Build #72886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/) succeeded d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9
* :broken_heart: [Build #69801](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/) failed 875d1d7cd052e64f711fd0b031936c4fbd760986
* :green_heart: [Build #69692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/) succeeded fe97375c19d92b02f98e5343901c12cdda1d5299
* :broken_heart: [Build #69681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69681/) failed 9b5fda21b5416dad6ea8ae2d1b294a3821b66e3c
* :broken_heart: [Build #69680](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69680/) failed cfe907bce365bb315959cdb53b8b0041507b89e4

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69680/","number":"69680","commit":"cfe907bce365bb315959cdb53b8b0041507b89e4"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69681/","number":"69681","commit":"9b5fda21b5416dad6ea8ae2d1b294a3821b66e3c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/","number":"69692","commit":"fe97375c19d92b02f98e5343901c12cdda1d5299"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/","number":"69801","commit":"875d1d7cd052e64f711fd0b031936c4fbd760986"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/","number":"72886","commit":"d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/","number":"73318","commit":"c2555e20ef6741cbc3f551d67472d93cf43525c8"}]}
PIPELINE-->

cchaos reviewed (approved):
Great, LGTM, thanks!

flash1293 commented:
@elasticmachine merge upstream

flash1293 commented:
@wylieconlon Are you fine with the formatting solution for the scope of this PR?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73607/)
* Commit: b0fcbf666bc5b8611a4250cfd8e7e4216434b907

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/b0fcbf666bc5b8611a4250cfd8e7e4216434b907) | diff | [baseline](https://github.com/elastic/kibana/commit/1dd4b65321570e8ce3df1b19a2fc7c85e6a4c25b) |
| --- | --- | --- | --- |
| lens | 258 | +3 | 255 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/b0fcbf666bc5b8611a4250cfd8e7e4216434b907) | diff | [baseline](https://github.com/elastic/kibana/commit/1dd4b65321570e8ce3df1b19a2fc7c85e6a4c25b) |
| --- | --- | --- | --- |
| lens | 889.9KB | +16.0KB | 874.0KB |

### History
* :green_heart: [Build #73318](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/) succeeded c2555e20ef6741cbc3f551d67472d93cf43525c8
* :green_heart: [Build #72886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/) succeeded d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9
* :broken_heart: [Build #69801](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/) failed 875d1d7cd052e64f711fd0b031936c4fbd760986
* :green_heart: [Build #69692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/) succeeded fe97375c19d92b02f98e5343901c12cdda1d5299
* :broken_heart: [Build #69681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69681/) failed 9b5fda21b5416dad6ea8ae2d1b294a3821b66e3c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69681/","number":"69681","commit":"9b5fda21b5416dad6ea8ae2d1b294a3821b66e3c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/","number":"69692","commit":"fe97375c19d92b02f98e5343901c12cdda1d5299"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/","number":"69801","commit":"875d1d7cd052e64f711fd0b031936c4fbd760986"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/","number":"72886","commit":"d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/","number":"73318","commit":"c2555e20ef6741cbc3f551d67472d93cf43525c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73607/","number":"73607","commit":"b0fcbf666bc5b8611a4250cfd8e7e4216434b907"}]}
PIPELINE-->

flash1293 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74018/)
* Commit: dac16610924508ecf1750ea9041e7f7079a5930b

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/dac16610924508ecf1750ea9041e7f7079a5930b) | diff | [baseline](https://github.com/elastic/kibana/commit/61951a577041317989140e22992b6c75410d34d9) |
| --- | --- | --- | --- |
| lens | 418 | +3 | 415 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/dac16610924508ecf1750ea9041e7f7079a5930b) | diff | [baseline](https://github.com/elastic/kibana/commit/61951a577041317989140e22992b6c75410d34d9) |
| --- | --- | --- | --- |
| lens | 1.0MB | +16.0KB | 1.0MB |

### History
* :green_heart: [Build #73607](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73607/) succeeded b0fcbf666bc5b8611a4250cfd8e7e4216434b907
* :green_heart: [Build #73318](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/) succeeded c2555e20ef6741cbc3f551d67472d93cf43525c8
* :green_heart: [Build #72886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/) succeeded d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9
* :broken_heart: [Build #69801](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/) failed 875d1d7cd052e64f711fd0b031936c4fbd760986
* :green_heart: [Build #69692](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/) succeeded fe97375c19d92b02f98e5343901c12cdda1d5299

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69692/","number":"69692","commit":"fe97375c19d92b02f98e5343901c12cdda1d5299"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/","number":"69801","commit":"875d1d7cd052e64f711fd0b031936c4fbd760986"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/","number":"72886","commit":"d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/","number":"73318","commit":"c2555e20ef6741cbc3f551d67472d93cf43525c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73607/","number":"73607","commit":"b0fcbf666bc5b8611a4250cfd8e7e4216434b907"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74018/","number":"74018","commit":"dac16610924508ecf1750ea9041e7f7079a5930b"}]}
PIPELINE-->

wylieconlon reviewed (approved):
LGTM, did not run the code again!

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74370/)
* Commit: dc040fe7c0c84b33e9c0ea189d5006053e5c7451

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/dc040fe7c0c84b33e9c0ea189d5006053e5c7451) | diff | [baseline](https://github.com/elastic/kibana/commit/948fad0b567bf951316bcf3b5569760f9769ea53) |
| --- | --- | --- | --- |
| lens | 436 | +3 | 433 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/dc040fe7c0c84b33e9c0ea189d5006053e5c7451) | diff | [baseline](https://github.com/elastic/kibana/commit/948fad0b567bf951316bcf3b5569760f9769ea53) |
| --- | --- | --- | --- |
| lens | 1.1MB | +16.0KB | 1.0MB |

### History
* :green_heart: [Build #74018](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74018/) succeeded dac16610924508ecf1750ea9041e7f7079a5930b
* :green_heart: [Build #73607](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73607/) succeeded b0fcbf666bc5b8611a4250cfd8e7e4216434b907
* :green_heart: [Build #73318](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/) succeeded c2555e20ef6741cbc3f551d67472d93cf43525c8
* :green_heart: [Build #72886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/) succeeded d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9
* :broken_heart: [Build #69801](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/) failed 875d1d7cd052e64f711fd0b031936c4fbd760986

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/69801/","number":"69801","commit":"875d1d7cd052e64f711fd0b031936c4fbd760986"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72886/","number":"72886","commit":"d8e2289bdcc7bbf3b49a9383551ebeb9e45521a9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73318/","number":"73318","commit":"c2555e20ef6741cbc3f551d67472d93cf43525c8"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/73607/","number":"73607","commit":"b0fcbf666bc5b8611a4250cfd8e7e4216434b907"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74018/","number":"74018","commit":"dac16610924508ecf1750ea9041e7f7079a5930b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74370/","number":"74370","commit":"dc040fe7c0c84b33e9c0ea189d5006053e5c7451"}]}
PIPELINE-->
